src/Symfony/Component/Config/Definition/JsonSchemaGenerator.php

```diff
@@ -0,0 +1,237 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Config\Definition;
+
+use Symfony\Component\Config\Definition\Builder\ExprBuilder;
+
+/**
+ * @experimental
+ */
+final readonly class JsonSchemaGenerator
+{
+    public function __construct(private string $outputPath)
+    {
+    }
+
+    public function merge(NodeInterface $node, \stdClass $schema = new \stdClass()): void
+    {
+        if (file_exists($this->outputPath)) {
+            $previousSchema = json_decode(file_get_contents($this->outputPath), flags: \JSON_THROW_ON_ERROR);
+            foreach ($previousSchema as $key => $value) {
+                $schema->$key ??= $value;
+            }
+        }
+
+        if (!\in_array($ref = '#/definitions/'.$node->getName(), array_column($schema->anyOf ??= [], '$ref'))) {
+            $schema->anyOf[] = (object) ['$ref' => $ref];
+        }
+
+        $this->build($node, $schema);
+    }
+
+    public function build(NodeInterface $node, \stdClass $schema = new \stdClass()): void
+    {
+        $rootNodeName = $node->getName() ?? 'root';
+
+        $schema->{'$schema'} ??= 'http://json-schema.org/draft-06/schema#';
+        if (!isset($schema->anyOf)) {
+            $schema->{'$ref'} ??= '#/definitions/'.$rootNodeName;
+        }
+        $schema->definitions ??= new \stdClass();
+        $schema->definitions->$rootNodeName = $this->buildSingleNode($node, $schema->definitions, allowParam: false);
+        $schema->definitions->param = (object) [
+            '$comment' => 'Container parameter',
+            'type' => 'string',
+            'pattern' => '^%[^%]+%$',
+        ];
+
+        file_put_contents($this->outputPath, json_encode($schema, \JSON_PRETTY_PRINT | \JSON_UNESCAPED_SLASHES | \JSON_THROW_ON_ERROR));
+    }
+
+    private function buildSingleNode(NodeInterface $node, \stdClass $definitions, bool $allowParam = true): \stdClass
+    {
+        $schema = (object) match (\count($types = $this->createSubSchemas($node, $definitions, $allowParam))) {
+            1 => $types[0],
+            default => ['anyOf' => $types],
+        };
+
+        if ($node->hasDefaultValue()) {
+            $schema->default = $node->getDefaultValue();
+        }
+
+        if ($node instanceof BaseNode) {
+            if ($info = $node->getInfo()) {
+                $schema->description = $info;
+            }
+
+            if ($node->isDeprecated()) {
+                $schema->deprecated = true;
+                $deprecation = $node->getDeprecation($node->getName(), $node->getPath());
+                $schema->deprecationMessage = ($deprecation['package'] || $deprecation['version'] ? "Since {$deprecation['package']} {$deprecation['version']}: " : '').$deprecation['message'];
```

Comment:
This can be simplified after rebase thanks to https://github.com/symfony/symfony/pull/61799
```suggestion
                $schema->deprecationMessage = $node->getDeprecationMessage();
```